captum/attr/_utils/lrp_rules.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class PropagationRule(ABC):
+    """
+    Base class for all propagation rule classes, also called Z-Rule.
+    STABILITY_FACTOR is used to assure that no zero divison occurs.
+    """
+
+    STABILITY_FACTOR = 1e-9
+
+    def forward_hook(self, module, inputs, outputs):
+        """Register backward hooks on input and output
+        tensors of linear layers in the model."""
+        self._handle_input_hooks = []
+        for input in inputs:
+            if not hasattr(input, "hook_registered"):
+                input_hook = self._create_backward_hook_input(input.data)
+                self._handle_input_hooks.append(input.register_hook(input_hook))
+                input.hook_registered = True
+        output_hook = self._create_backward_hook_output(outputs.data)
```

Comment:
nit: usually it is not the case but the output could be a tuple of tensors as well ?

Reply:
So far I only observed that the `inputs` of the forward hook were always tuple even if it was of length 1 and outputs were always tensors. But I agree that it could also be tuples in general. Then, I may add a case for outputs of type tuple?

Reply:
yes, it is not common that the output of a layer is a tuple but it can sometimes happen. We support those cases for other attribution algorithms as well. It would be good to be consistent and support it in LRP as well.
We might have such a test model with a test layer that returns a tuple. Let me see if I can find it and I'll point it out to you.

Reply:
I guess if we assume that the layers are always standard layers such as a linear layer then we can assume that the output is always a tensor but if it is a custom layer then it can potentially return a tuple.